### DISCUSSION TITLE
useFetcher data is set to undefined, but only on the first load().

### DISCUSSION THREAD

**Discussion by ComeToMiCasa** [remix-run/remix#7566] [Q&A]
I'm trying to use `useFetcher()` from `@remix-run/react` to fetch data from the client side when the client clicks a button. However, `fetcher.data` is set to `undefined` but only on the first `load()`. 

my `app/routes/api.test.tsx`:

```js
export const loader = async ({ request }: LoaderFunctionArgs) => {
	return {
		data: "test"
	};
};
```

my `app/routes/_index.tsx/`:
```js
export default function Index() {

	const testFetcher = useFetcher();


	const handleClick = () => {
		testFetcher.load("/api/test");
		console.log(testFetcher);
	};

	return (
		<div >
			<button onClick={handleClick}>click</button>
		</div>
	);
}

```

and my `console` after two clicks of the button:
[![enter image description here][1]][1]

I first thought it was because `fetcher.load()` is asynchronous and the fetching wasn't done yet, but adding async/await didn't work because it wasn't an asynchronous function after all. 

Also, if I manually fetch the data by entering `http://localhost:3000/api/test` in my browser, I can clearly see the data. What am I doing wrong? 

  [1]: https://i.stack.imgur.com/qIbIR.png

**Comment by kiliman** [comment]
Fetchers go through multiple states. From idle -> loading -> back to idle with data. If you console log the fetcher state and data, you'll see what's happening. 

Just check to see if data is populated before using it.

**Reply by ComeToMiCasa** [reply]
If the the reason why data is undefined on the first log in the image is because the data is yet to be populated, shouldn't the state be `loading`, not `idle`? 

Also, wouldn't it be a better practice to make the `load()` function a promise?

**Reply by sergiodxa** [reply]
> If the the reason why data is undefined on the first log in the image is because the data is yet to be populated, shouldn't the state be `loading`, not `idle`?

You're thinking about useFetcher as useQuery from RQ I think, while it can be used to get data it can also be used to post data to the server, to as a fetcher.Form to both get or post data. Think of it as a [lazy useQuery](https://tanstack.com/query/v4/docs/react/guides/disabling-queries#lazy-queries).

So the reason it doesn't start with `state` being `loading` is because until you call `fetcher.load` or `fetcher.submit` or a `fetcher.Form` is submitted it will not do anything.

> Also, wouldn't it be a better practice to make the `load()` function a promise?

This has been requested in the past, Ryan has answered it in other discussions why you should use the state and data values instead of awaiting a promise
